packages/ember-htmlbars/lib/hooks/create-fresh-scope.js

```diff
@@ -46,13 +49,124 @@
   the current view's `controller`.
 */
 
+function Scope(parent) {
+  this._self = null;
+  this._blocks = null;
+  this._component = null;
+  this._view = null;
+  this._attrs = null;
+  this._locals = null;
+  this._localPresent = null;
+  this.overrideController = null;
+  this.parent = parent;
+}
+
+let proto = Scope.prototype;
+
+proto.getSelf = function() {
+  return this._self || this.parent.getSelf();
+};
+
+proto.bindSelf = function(self) {
+  this._self = self;
+};
+
+proto.updateSelf = function(self, key) {
+  let existing = this._self;
+
+  if (existing) {
+    existing.setSource(self);
+  } else {
+    this._self = new ProxyStream(self, key);
+  }
+};
+
+proto.getBlock = function(name) {
+  if (!this._blocks) { return this.parent.getBlock(name); }
+  return this._blocks[name] || this.parent.getBlock(name);
+};
+
+proto.hasBlock = function(name) {
+  if (!this._blocks) { return this.parent.hasBlock(name); }
+  return !!(this._blocks[name] || this.parent.hasBlock(name));
+};
+
+proto.bindBlock = function(name, block) {
+  if (!this._blocks) { this._blocks = new EmptyObject(); }
+  this._blocks[name] = block;
+};
+
+proto.getComponent = function() {
+  return this._component || this.parent.getComponent();
+};
+
+proto.bindComponent = function(component) {
+  this._component = component;
+};
+
+proto.getView = function() {
+  return this._view || this.parent.getView();
+};
+
+proto.bindView = function(view) {
+  this._view = view;
+};
+
+proto.getAttrs = function() {
+  return this._attrs || this.parent.getAttrs();
+};
+
+proto.bindAttrs = function(attrs) {
+  this._attrs = attrs;
+};
+
+proto.hasLocal = function(name) {
+  if (!this._localPresent) { return this.parent.hasLocal(name); }
+  return this._localPresent[name] || this.parent.hasLocal(name);
+};
+
+proto.hasOwnLocal = function(name) {
+  return this._localPresent && this._localPresent[name];
+};
+
+proto.getLocal = function(name) {
+  if (!this._localPresent) { return this.parent.getLocal(name); }
+  return this._localPresent[name] ? this._locals[name] : this.parent.getLocal(name);
+};
+
+proto.bindLocal = function(name, value) {
+  if (!this._localPresent) {
+    this._localPresent = new EmptyObject();
+    this._locals = new EmptyObject();
+  }
+
+  this._localPresent[name] = true;
+  this._locals[name] = value;
+};
+
+const EMPTY = {
```

Comment:
Fixed in https://github.com/emberjs/ember.js/commit/742cbaa9094361c7e9938bba969cbaf8be778853.